src/token/TokenUpgrade.sol

```diff
@@ -96,26 +103,37 @@ contract TokenSwap is AccessControl() {
     /// @param token The token to be recovered
     /// @param to The address to send the tokens to
     function recover(IERC20 token, address to) external auth {
-        require(address(tokensIn[token].reverse) == address(0), "TokenIn registered");
-        require(address(tokensOut[token].reverse) == address(0), "TokenOut registered");
+        if (address(tokensIn[token].reverse) != address(0)) revert TokenInAlreadyRegistered(address(token));
+        if (address(tokensOut[token].reverse) != address(0)) revert TokenOutAlreadyRegistered(address(token));
         uint256 recovered = token.balanceOf(address(this));
         token.safeTransfer(to, token.balanceOf(address(this)));
 
         emit Recovered(token, recovered);
     }
 
-    /// @dev Swap a token for its replacement, at the registered ratio. The tokens must have been sent to the contract before this call.
+    /// @dev Swap a token for its replacement, at the registered ratio.
     /// @param tokenIn_ The token to be replaced
-    function swap(IERC20 tokenIn_, address to) external {
+    /// @param from the owner of tokenIn_
+    /// @param to the receiver of tokenOut_
+    /// @param tokenInAmount The amount of tokenIn_ to swap
+    /// @param proof The merkle proof to verify the swap
+    function swap(IERC20 tokenIn_, address from, address to, uint256 tokenInAmount, bytes32[] calldata proof)
+        external
+    {
         TokenIn memory tokenIn = tokensIn[tokenIn_];
-        require(address(tokenIn.reverse) != address(0), "TokenIn not registered");
+        if (address(tokenIn.reverse) == address(0)) revert TokenInNotRegistered(address(tokenIn_));
         IERC20 tokenOut_ = tokenIn.reverse;
 
-        uint256 tokenInAmount = tokenIn_.balanceOf(address(this)) - tokenIn.balance;
+        tokenIn_.safeTransferFrom(from, address(this), tokenInAmount);
         uint256 tokenOutAmount = tokenInAmount * tokenIn.ratio / 1e18;
```

Comment:
```suggestion
        uint256 tokenOutAmount = tokenInAmount.wmul(tokenIn.ratio);
```

Would the rounding mean that the TokenUpgrade contract gets the left over wei, or that there is a shortfall of a few wei if all the users swap their tokens? Worth noting as natspec.